# Exercise 02 -  OLAP Cubes - Slicing and Dicing

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by creating and connecting to the database by running the cells below.

In [1]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

 set_config 
------------
 
(1 row)

 setval 
--------
    200
(1 row)

 setval 
--------
    605
(1 row)

 setval 
--------
     16
(1 row)

 setval 
--------
    600
(1 row)

 setval 
--------
    109
(1 row)

 setval 
--------
    599
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
   1000
(1 row)

 setval 
--------
   4581
(1 row)

 setval 
--------
      6
(1 row)

 setval 
--------
  32098
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
      2
(1 row)

 setval 
--------
      2
(1 row)



### Connect to the local database where Pagila is loaded

In [2]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

# Start with a simple cube
TODO: Write a query that calculates the revenue (sales_amount) by day, rating, and city. Remember to join with the appropriate dimension tables to replace the keys with the dimension labels. Sort by revenue in descending order and limit to the first 20 rows. The first few rows of your output should match the table below.

In [3]:
%%time
%%sql

SELECT dimdate.day, dimmovie.rating, sum(factsales.sales_amount) as revenue, dimcustomer.city
FROM factSales
JOIN dimdate on (dimdate.date_key = factsales.date_key)
JOIN dimmovie on(dimmovie.movie_key = factsales.movie_key)
JOIN dimcustomer on (dimcustomer.customer_key = factsales.customer_key) 
group by (dimdate.day, dimmovie.rating, dimcustomer.city)
order by revenue desc
limit 5;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
5 rows affected.
CPU times: user 5.01 ms, sys: 44 µs, total: 5.06 ms
Wall time: 72.4 ms


[(30, 'G', Decimal('24.97'), 'San Bernardino'),
 (30, 'NC-17', Decimal('23.95'), 'Apeldoorn'),
 (21, 'NC-17', Decimal('22.97'), 'Belm'),
 (30, 'PG-13', Decimal('21.97'), 'Zanzibar'),
 (28, 'R', Decimal('21.97'), 'Mwanza')]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>San Bernardino</td>
        <td>24.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>Apeldoorn</td>
        <td>23.95</td>
    </tr>
    <tr>
        <td>21</td>
        <td>NC-17</td>
        <td>Belm</td>
        <td>22.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>R</td>
        <td>Mwanza</td>
        <td>21.97</td>
    </tr>
</tbody></table></div>

## Slicing

Slicing is the reduction of the dimensionality of a cube by 1 e.g. 3 dimensions to 2, fixing one of the dimensions to a single value. In the example above, we have a 3-dimensional cube on day, rating, and country.

TODO: Write a query that reduces the dimensionality of the above example by limiting the results to only include movies with a `rating` of "PG-13". Again, sort by revenue in descending order and limit to the first 20 rows. The first few rows of your output should match the table below. 

In [8]:
%%time
%%sql

SELECT dimdate.day, dimmovie.rating, sum(factsales.sales_amount) as revenue, dimcustomer.city
FROM factSales
JOIN dimdate on (dimdate.date_key = factsales.date_key)
JOIN dimmovie on(dimmovie.movie_key = factsales.movie_key)
JOIN dimcustomer on (dimcustomer.customer_key = factsales.customer_key) 
WHERE dimmovie.rating in ('PG-13') 
group by (dimdate.day, dimmovie.rating, dimcustomer.city)
order by revenue desc
limit 5;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
5 rows affected.
CPU times: user 4.92 ms, sys: 0 ns, total: 4.92 ms
Wall time: 16 ms


[(30, 'PG-13', Decimal('21.97'), 'Zanzibar'),
 (28, 'PG-13', Decimal('19.97'), 'Dhaka'),
 (30, 'PG-13', Decimal('18.97'), 'Osmaniye'),
 (29, 'PG-13', Decimal('18.97'), 'Shimoga'),
 (21, 'PG-13', Decimal('18.95'), 'Asuncin')]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>PG-13</td>
        <td>Dhaka</td>
        <td>19.97</td>
    </tr>
    <tr>
        <td>29</td>
        <td>PG-13</td>
        <td>Shimoga</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Osmaniye</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>21</td>
        <td>PG-13</td>
        <td>Asuncin</td>
        <td>18.95</td>
    </tr>
</tbody></table></div>

## Dicing
Dicing is creating a subcube with the same dimensionality but fewer values for  two or more dimensions. 

TODO: Write a query to create a subcube of the initial cube that includes moves with:
* ratings of PG or PG-13
* in the city of Bellevue or Lancaster
* day equal to 1, 15, or 30

The first few rows of your output should match the table below. 

In [10]:
%%time
%%sql

SELECT dimdate.day, dimmovie.rating, sum(factsales.sales_amount) as revenue, dimcustomer.city
FROM factSales
JOIN dimdate on (dimdate.date_key = factsales.date_key)
JOIN dimmovie on(dimmovie.movie_key = factsales.movie_key)
JOIN dimcustomer on (dimcustomer.customer_key = factsales.customer_key) 
WHERE dimmovie.rating in ('PG-13', 'PG') 
AND dimcustomer.city in ('Bellevue', 'Lancaster')
AND dimdate.day in ('1', '15', '30')
group by (dimdate.day, dimmovie.rating, dimcustomer.city)
order by revenue desc
limit 5;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
5 rows affected.
CPU times: user 5.15 ms, sys: 267 µs, total: 5.42 ms
Wall time: 10.2 ms


[(30, 'PG', Decimal('12.98'), 'Lancaster'),
 (1, 'PG-13', Decimal('5.99'), 'Lancaster'),
 (30, 'PG-13', Decimal('3.99'), 'Bellevue'),
 (30, 'PG-13', Decimal('2.99'), 'Lancaster'),
 (15, 'PG-13', Decimal('1.98'), 'Bellevue')]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>Lancaster</td>
        <td>12.98</td>
    </tr>
    <tr>
        <td>1</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>5.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>3.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>2.99</td>
    </tr>
    <tr>
        <td>15</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>1.98</td>
    </tr>
</tbody></table></div>